### PR TITLE
ENH: Add class to write dta format 117 files

### LINKED ISSUE #16450: to_stata: Fixed width strings in Stata .dta files are limited to 244 (or fewer)
#### Code Sample, a copy-pastable example if possible

```python
import pandas
frame = pandas.DataFrame({'A':['h'*250,'hi','hola']})
frame.to_excel("text.xlsx", index=False)
frame.to_stata("test.dta")

```
#### Problem description

Raises the following error: 
> ValueError:
> Fixed width strings in Stata .dta files are limited to 244 (or fewer)
> characters.  Column 'A' does not satisfy this restriction.

However this restriction seems to not exists in STATA, as the Excel file can be imported correctly
Open STATA, import the Excel file

```STATA
import excel "C:\data\tesi\software\text.xlsx", sheet("Sheet1") firstrow clear

```
Now we can get the type of data in column 'A', and as you can see, it's str250. So STATA can store string longer than 244 characters

```STATA
. describe A

```
A               str250  %250s                 A

#### Expected Output
File gets exported with the correct format and without problems

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 23.0.0
Cython: 0.24
numpy: 1.11.2
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 4.2.0
sphinx: 1.5.5
patsy: None
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: None
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.7.3
bs4: 4.5.3
html5lib: 0.999999999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: 0.2.1
</details>

### PR DESCRIPTION
**Pull request by bashtage**
Add export for dta 117 files which add support for long strings
Refactor StataWriter to simplify new writer

closes #16450

- [x] closes #16450
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by bashtage** [comment]
Is it intentional that `stata.py` is not covered?  Seems a bit strange.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
...
+            read by Stata 10 and later.  Version 117 can be read by Stata 13
+            or later. Version 114 limits string variables to 244 characters or
+            fewer while 117 allows strings with lengths up to 2,000,000
+            characters.
+
+            .. versionadded:: 0.23.0
+
+        convert_strl : list, optional
+            List of column names to convert to string columns to Stata StrL
+            format. Only available if version is 117.  Storign strings in the
```
Comment (jreback) [inline pandas/core/frame.py]: Storing

**Comment by jreback** [comment]
> Is it intentional that stata.py is not covered? Seems a bit strange.

what does this mean? the coverage that prints on the issue is odd (and has always been that way), but you can click thru and see the actual reports.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
 
             .. versionadded:: 0.19.0
 
+        version : {114, 117}
+            dta version to use in the output file.  Version 114 can be used
+            read by Stata 10 and later.  Version 117 can be read by Stata 13
```
Comment (jreback) [inline pandas/core/frame.py]: should we just support version 117 and forward? how old are these respective versions?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (jreback) [reply pandas/core/frame.py]: how much simplfication do we get from only supporting new format?

File: pandas/io/stata.py [hunk 3]
```diff
@@ -2307,3 +2384,553 @@ def _null_terminate(self, s, as_string=False):
...
+        """Write the file header"""
+        byteorder = self._byteorder
+        self._file.write(_bytes('<stata_dta>', 'utf-8'))
+        bio = BytesIO()
+        # ds_format - 117
+        bio.write(self._tag(_bytes('117', 'utf-8'), 'release'))
+        # byteorder
+        bio.write(self._tag(byteorder == ">" and "MSF" or "LSF", 'byteorder'))
+        # number of vars, 2 bytes
+        assert self.nvar < 2 ** 16
```
Comment (jreback) [inline pandas/io/stata.py]: maybe add a blank line before comments, easier to read

File: pandas/tests/io/test_stata.py [hunk 4]
```diff
@@ -5,21 +5,22 @@
 import os
 import struct
 import warnings
-from datetime import datetime
 from collections import OrderedDict
+from datetime import datetime
 
 import numpy as np
+import pytest
+from pandas._libs.tslib import NaT
```
Comment (jreback) [inline pandas/tests/io/test_stata.py]: import from pandas

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments. Might consiser renaming StataWriter to StataWriter114 (and essentially make it private). Otherwise lgtm.

**Comment by bashtage** [comment]
> minor comments. Might consiser renaming StataWriter to StataWriter114 (and essentially make it private). Otherwise lgtm.

I think this would need a dep cycle since `StataWriter` has been public.  Can do this if you want.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
 
             .. versionadded:: 0.19.0
 
+        version : {114, 117}
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Hmm, should these be strings? Is this exposed anywhere in stats itself? Do they use integers? (when I see version number, I think string).

Comment (TomAugspurger) [reply pandas/core/frame.py]: Huh, https://www.stata.com/support/faqs/data-management/save-for-previous-version/ seems to suggest that stata uses integers? `version(13)`. OK then, let's follow that.

File: pandas/core/frame.py [hunk 6]
```diff
@@ -1832,12 +1856,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
         >>> writer = StataWriter('./date_data_file.dta', data, {2 : 'tw'})
         >>> writer.write_file()
         """
-        from pandas.io.stata import StataWriter
-        writer = StataWriter(fname, self, convert_dates=convert_dates,
+        kwargs = {}
+        if version not in (114, 117):
+            raise ValueError('Only formats 114 and 117 supported.')
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Would be nice to include the user passed `version` in the error message.

File: pandas/tests/io/test_stata.py [hunk 7]
```diff
@@ -1355,3 +1413,63 @@ def test_date_parsing_ignores_format_details(self, column):
...
+                                  original, check_index_type=False)
+            tm.assert_frame_equal(original, copy)
+
+    def test_convert_strl_name_swap(self):
+        original = DataFrame([['a' * 3000, 'A', 'apple'],
+                              ['b' * 1000, 'B', 'banana']],
+                             columns=['long1' * 10, 'long', 1])
+        original.index.name = 'index'
+
+        with warnings.catch_warnings(record=True) as w:  # noqa
```
Comment (TomAugspurger) [inline pandas/tests/io/test_stata.py]: What warnings are you catching here?

**Review by TomAugspurger — APPROVED** [review]
Couple small comments. OK to do as a followup (I'm tagging the RC momentarily.)

**Comment by TomAugspurger** [comment]
@jreback are you OK with merging this as is? I'll fix the git conflict on merge.

**Comment by jreback** [comment]
yep this is fine
if we need follow ups after RC that is fine

**Comment by TomAugspurger** [comment]
K. @bashtage I'm going to fix the conflict and then merge. A followup PR would be welcome.

**Pull request merged**